portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -195,7 +251,92 @@ private String getImageKey(HttpServletRequest request, String formElementID) {
     
     // In previous versions, ImagesServices was used to get a URL pointing to the uploaded img.
     // Here, we simpy return the Blob's key so that it can later be served directly.
-    return blobKey.getKeyString();
+    return blobKey;
+  }
+
+  /** 
+   * Blobstore stores files as binary data; this func retrieves the data stored at imageKey. 
+   * (See image-analysis directory example)
+   */
+  private byte[] getBlobBytes(BlobKey imageKey) throws IOException {
+    BlobstoreService bs = BlobstoreServiceFactory.getBlobstoreService();
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currByteIndex = 0;
+    boolean continueRead = true;
+    while(continueRead) {
+      // Fetch a portion of the image bytes from Blobstore.
+      byte[] temp = bs.fetchData(imageKey, currByteIndex, currByteIndex + fetchSize - 1); 
+      outputBytes.write(temp);
+
+      // If fewer bytes than requested are read, then the end was reached.
+      if (temp.length < fetchSize) {
+        continueRead = false;
+      }
+
+      currByteIndex += fetchSize;
+    }
+    return outputBytes.toByteArray();
+  }
+
+
+  /** 
+   * Uses Cloud Vision API to generate a list of objects for the image
+   * represented by the binary data stored in the imgBytes param.
+   */
+  private List<LocalizedObjectAnnotation> getImageObjects(byte[] imageBytes) throws IOException {
+    ByteString byteString = ByteString.copyFrom(imageBytes);
+    Image image = Image.newBuilder().setContent(byteString).build();
+
+    // Use a Feature to make an annotate-image request; add this request to a list of such.
+    Feature feature = Feature.newBuilder().setType(Feature.Type.OBJECT_LOCALIZATION).build();
+    AnnotateImageRequest request = 
+        AnnotateImageRequest.newBuilder().addFeatures(feature).setImage(image).build();
+    List<AnnotateImageRequest> requests = new ArrayList<>();
+    requests.add(request);
+
+    // Batch annotate all the image requests in the list (only one, here) and get the responses.
+    ImageAnnotatorClient client = ImageAnnotatorClient.create();
+    BatchAnnotateImagesResponse batchResponse = client.batchAnnotateImages(requests);
+    client.close();    // client should always be closed
+    AnnotateImageResponse imageResponse = batchResponse.getResponsesList().get(0);
+ 
+    if (imageResponse.hasError()) {
```

Comment:
Do you have to return `null` here? Look like having this set doesn't [necessarily mean there's no annotation](https://googleapis.dev/java/google-cloud-vision/latest/com/google/cloud/vision/v1/AnnotateImageResponse.html#hasError--).

Reply:
I think it's worth returning null since it's not exactly clear what the annotation looks like in this case; this also makes it easier to test up above if there are image labels or not, just test == null.